### PR TITLE
S^4: first cut at a working scheduler

### PR DESCRIPTION
**Pull request by eteq**
This PR implements s<sup>4</sup>: the Stupid Simple Sequential Scheduler.  The basic algorithm is simply this: given a list of targets/observing blocks, a time window, and a set of constraints, step through the time window, assigning observations to whatever the best target is that hasn't yet been observed. You can see the essential functionality by running the ``astroplan/tests/s4_experiments.ipynb`` notebook, or you can just [look at how it the output from when I ran it](https://github.com/eteq/astroplan/blob/staging/astroplan/tests/s4_experiments.ipynb).

Note that, perhaps more importantly than the scheduler itself, this PR also adds some of the scaffolding necessary to build schedulers.  Specifically, that includes:
* `ObservingBlock`: represents a chunk of time spent observing a target, possibly also including information like what instrument is used, what filter needs to be in, etc. The user specifies the inputs (like exposure time, filter, etc.), but then the scheduler populates the actual specific time windows.
* `TransitionBlock`: a chunk of time spent transitioning between observing blocks.  Includes info on why the transition happened (telescope had to slew, filter had to change, etc.)
* `Transitioner`: a class that can describe in a general way how to transition between blocks.  Specific instances could be provided e.g. by an observatory so that users can just treat it as a black-box.

The other significant thing it does is make a subtle but important change to the constraints classes.  Previously, constraints were always boolean.  That is, either you met the constraint, or not.  This PR relaxes that a bit and says constraints can also be floats on the range [0,1].  They are then interpreted by the scheduler as priorities.  E.g., `AboveAirmass` now smoothly transitions from 0 to 1 as you go from the max to min airmass.  So anything above the max still isn't observable (anything with a priority of 0 is not scheduled), but objects that are between min and max are prioritized higher for being at low airmass.  The scheduler can then use that to decide the "best" target.

This PR is not really ready to be merged in its current state: it has no real tests, nor documentation (and even some of the docstrings need addition/cleanup).  But before spending more time on this, I thought I should put it out and see what folks think.

cc bmorris3 @cdeil @adrn @ejeschke @jberlanga

### CODE REVIEW

**Comment by ejeschke** [comment]
An excellent start!

In our queue scheduler for Subaru we also ended up using weights to evaluate the OBs. The weights are great for things like program ranks, time to slew to a target etc, whether you have to take time to change a filter, etc.  But we also kept boolean constraints as well, because there are some things that just block an OB completely, like not having a certain filter installed, or the target not being visible at all, etc.  We do a first pass to remove all OBs that can't execute due to the absolute constraints, and then evaluate the others based on the sum of the weights.  This won't matter much unless you have many OBs to schedule over many nights.

I guess my main hope for this kind of PR would be to keep the OB structure as flexible as possible, since the kind of information stored in them will vary widely.  We use a generic OB with 4 sub-structures: a telescope configuration, an instrument configuration, a target and an environment configuration (desired seeing, transparency, airmass, etc.)

**Comment by bsipocz** [comment]
@eteq - I wonder whether it would make some sense to include some global goodness measure, too. e.g. minimizing the airmass sum of all the targets observed or similar, so you don't start with the best one that will be also quite good for the whole time while the rest deteriorate faster.

**Comment by bmorris3** [comment]
This looks like a great start. @ejeschke - do you have an handle for how this compares to simple canonical schedulers? 

Maybe we should start reviewing that "literature" we've always talked about and see if the classes you've started here are generic enough to support the more complex schedulers we may need to work with later.

File: astroplan/constraints.py [hunk 1]
```diff
@@ -635,3 +692,11 @@ def observability_table(constraints, observer, targets, times=None,
     tab.meta['constraints'] = constraints
 
     return tab
+
+def _rescale_minmax(vals, min_val, max_val):
```
Comment (bmorris3) [inline astroplan/constraints.py]: @eteq - so all constraints will now have an additional option to output a float which "rates" the observability of a target on `(0, 1)`? I don't know if I understand what's going on with the maximum value here -- if the value is above the `max`, the rescaled value is `1` (not to be confused with `True`)?

**Comment by cdeil** [comment]
@eteq – This is exciting, thanks!

I don't have time to look at this in the near future, sorry. Just wanted to say that I don't think we have to get it right the first time ... if we put a big fat warning at the top that the API is experimental and feedback / improvements welcome, I think it would be best to merge this soon and even try to get many people to use this and give feedback by releasing and advertising it in a 0.2 release before the end of the year.

**Comment by bmorris3** [comment]
This is an opportunistic, impatient scheduler. I don't think/know if it has this behavior here, but I think it would be good to be clear whether or not the scheduler is choosing targets with knowledge of where they'll be at all of the blocks. It seems like your example run chooses the object at the lowest airmass first, then crosses it off the list, and in the next observing block chooses the target with the next lowest airmass. 

Wouldn't it be more generically useful to say: "I need to observe all four of these stars within this time window, and I'd like to observe each of them when it's most opportune to do so". For this example we would observe Vega first, since it's always at the highest airmass of the targets but starts at the lowest airmass that it will have during the time window, then Altair and so on, so that each target is observed at the lowest airmass that it can be observed within the available time. In this case we'd be taking the maximum of the product of the times and constraints of a target per block compared to the other potential blocks for that target.

Of course depending on context, both what I'm suggesting and what you wrote are useful. If the weather is questionable, I'd go with what you wrote -- give me the sure bet now, and we'll see if we can get the lower S/N targets later, for example.

**Comment by joshwalawender** [comment]
If one chooses a target by summing the output of the constraints (weighted I hope), then @bmorris3, if I wanted to get the behavior you mentioned, I would write the constraints to maximize at the most opportune time and you would get that behavior.

We seem to be discussing this as a tool to set a schedule for the night, but the more appealing case to me is a dispatch scheduler which chooses the best target at that particular time, then repeats the process after that observation is done.  There's a good discussion of this in [Denny (2004)](http://adsabs.harvard.edu/abs/2004SASS...23...35D) and we planned to use something [similar](https://github.com/panoptes/POCS/wiki/Scheduler-Outline) in project [PANOPTES](http://panoptes.github.io).

**Comment by bmorris3** [comment]
Hi @joshwalawender, thanks for chiming in! It's great to have your input. 

We're trying to build tools that are general enough for the night-long or longer use cases that @eteq and I were talking about, but also cases like yours. Since @eteq's scheduler takes a start and end time parameter to schedule between, could you run this scheduler over short windows from now until ~1 observation from now to get the behavior you're looking for?

**Comment by joshwalawender** [comment]
Yeah, this system running whenever a new target is needed should work for us no problem.  I've been meaning to rewrite the POCS scheduler to use astropy coordinates (instead of pyephem), but now it looks like my rewrite is down to ```import astroplan```. :)

I guess what I was trying to get at was that I think one can do the night-long use cases by writing appropriate custom constraints, right?  If the constraint is very peaky around the optimum time (determined by whatever your science/operational goal is), it should get selected at that time as long as there are other targets to do before then.  If the target list is too short to fill the time, then it would get observed beforehand.

**Comment by bmorris3** [comment]
I'm glad to hear that astroplan may be useful to you :smile: 

> I guess what I was trying to get at was that I think one can do the night-long use cases by writing appropriate custom constraints, right?

I think the constraints that you'd need are already written (`AirmassConstraint`, `AltitudeConstraint`, `MoonSeparationConstraint` etc), but the functionality that @eteq implemented here - what we're calling the "scheduler" - is just pulling out the "most observable" object at each observing block, then moving to the next "most observable" object in the next observing block. In that way, it sounds like what you need for a Dispatch Scheduler. 

>  If the constraint is very peaky around the optimum time (determined by whatever your science/operational goal is), it should get selected at that time as long as there are other targets to do before then.

I think see what you're saying -- if you were to iterate @eteq's implementation over small blocks, you'd get the kind of behavior I was asking for [above](https://github.com/astropy/astroplan/pull/137#issuecomment-155543338).

**Comment by joshwalawender** [comment]
I think I see now that one way to generate what you describe would be an optimization on some merit value for the full night schedule (rather than for the target), so you'd iterate over schedules and pick the best schedule.  That's potentially different than generating a single schedule based on potentially complex constraints which is what I was thinking.

**Comment by joshwalawender** [comment]
Hi @eteq, I have a question about ObservingBlocks.  It seems that in your definition that the ObservingBlock is the "minimum schedulable unit" (i.e. it is the unit thing that the scheduler places in time), but what if I need to link ObservingBlocks?  For example, if I want to observe Vega in the B filter, then immediately in the V filter to get color?  Those would each be ObservingBlocks with a TransitionBlock in between them to reconfigure the filter wheel, but the scheduler doesn't know that they have to be sequential in time.  Is there a good way to represent that?

I'm thinking that there might be another object: either a Minimum Schedulable Block (MSB) or a visit in various terminologies that I've encountered.  An MSB would basically be a sequential list of ObservingBlocks and TransitionBlocks and the MSB would be the unit thing that the scheduler should sequence.

One might also imagine an MSB which included multiple targets.  For example one might want to have the following sequence as an MSB:

* observe target in B (an ObservingBlock)
* transition to V filter (a TransitionBlock)
* observe target in V (an ObservingBlock)
* slew to a standard star (a TransitionBlock)
* observe standard in V (an ObservingBlock)
* transition to B filter (a TransitionBlock)
* observe standard in B (an ObservingBlock)

What do you think?  Do we need something to represent this or can we do it within the existing structure?

**Comment by joshwalawender** [comment]
@eteq I also see that S^4 determines the "merit" of a particular ObservingBlock by the product of all constraint results.  In Denny 2004 section 8 he uses a weighted sum where each constraint has it's own configurable weighting factor.  Denny also references Steele and Carter 1997 (who discuss the 2m Liverpool Telescope) as using the same basic idea.

The advantage of the product is that a constraint result of 0 vetoes the ObservingBlock by definition (that behavior would have to be enforced in another way if we used a sum), but I'm not sure that using a product would generate the behavior I expect from the scheduler.  For example, if one of many constraints yielded a very low score it would strongly depress the ObservingBlock's merit where in a sum, one low score (but not a veto) would not depress the merit quite as strongly.

**Comment by eteq** [comment]
Thanks for all the quick feedback!  Just so everyone doesn't think I'm ignoring them, I haven't responded to most of this yet because I was on the observing run that inspired me to implement this PR... I'll try to address some of the big items now, but might have more questions/comments later after I've mulled this over a bit more.

@ejeschke - Yes, completely agree with you that the OB should be a very flexible object.  Right now the class has telescope/instrument configuration basically lumped into a single "configuration" dict - So users can say things like "it takes 2 minutes to switch from the r to g band filter" and they encode that as tuples and time `Quantities` in a nested dict. That could be expanded to be easier to work with if we want, but that seemed the most straightforward place to start. The "environment configuration" is a bit different here, though, as I guess it's sort of encoded in the constraints.  But maybe that's not good?  It seemed natural from what we already had, but that's not necessarily a good argument...

@bsipocz - what do you mean by "global goodness measure"?  You mean like a user rating along the lines of "I think this is more important than that"?  Right now there is a global goodness derived from the product of all of the constraints' goodness, but maybe that's not what you meant?

@bmorris3 - yes, you're 100% right about the strategy here, and I agree it is not at all optimal for many cases.  It's best for the case where you have way more targets than you can observe and just want to grab the best at any given moment.  I wrote it this way purely because it's the simplest thing I could imagine to implement on the actual scheduling side (hence the "stupid").  We definitely want to give users the option for *both*, but I'm hoping that we can use this to develop the *framework* of `ObservingBlocks` and such, and then in a follow-on, someone can actually start working on more complex scheduling strategies.

That's why the S4 scheduler is called `SequentialScheduler` and a distinct class - the ideal is that you could do ``SequentialScheduler(...)`` when you need this use case, and ``BestNightScheduler(...)`` when you wanted the other case, and probably others we haven't thought of, but then you'd feed them observing blocks, constraints, and invoke them all in the same way.

@joshwalawender - re: OBs: My thinking when writing this is that you do *not* want to have to micromanage your scheduler.  If you tell it "I need to have V right after B", that's actually bad - it might be that for various reasons it's more optimal to get Vega in V, then go to something else in V, then come back to Vega in B later (I'm thinking of a conversation I had with some LSST folks, where filter changes are relatively expensive in terms of dead-time).  So you want your scheduler to make that decision for you.  

That said, there are reasons why you might want to have specific linked constraints like "I need to observe this variable star ata cadence of once every minute for an hour, otherwise it's useless to observe it at all".  That's not covered at all right now.  I could see that addressed in two possible ways: 1) An MSB along the lines of what you're suggesting above 2) A constraint that says "if they aren't 20 blocks of this target next to each other, the constraint fails".  2 Is more "pure", but 1 is probably more practical. My prefernce would be to not implement that sort of thing in this PR, rather as a follow-on, but it's definitely something we should have in mind as a use case.

One other note about this: it may not be clear from the description, but in the current design, you never want to have a user create a `TransitionBlock` - that's only an *output* of the scheduler.  So even if we implement the MSB, I'd say you write it as a sequence of `ObservingBlock`s, and the scheduler has the duty of creating all the transitions.  That way you can do things like comparing the schedulability of two different telescopes (with different transition times) with the same set of observing blocks.

@joshwalawender - re: merit/weighting approach - I see your point (And I'm probably much less in contact with the scheduling literature than you are).  But this raises the question: do we want to allow multiple ways of deciding how to combine weights/constraints?  I.e., a `WeightCombiner` class or something?  It could default to whatever seems best, but the user could then choose from a builtin set or write their own if they want more control...?


General question for anyone who's interested: Is it a Good Thing (TM) to be treating `Constraint`s as "weights"?  Or should we have a separate heirarchy of things called `Weight`s that map to [0,1], and reserve `Constraint`s for True/False?  I didn't think of this at the time, but the discussion above makes me think that might be more conceptually easy to follow?

**Comment by joshwalawender** [comment]
Regarding the question of splitting ```Weight```s and ```Constraint```s, I could see the use in that.  We took a different tack when drafting the PANOPTES scheduler and had each "merit function" (in our terminology) return a tuple which has a float value (the merit of that OB according to that merit function) and a boolean (which indicated whether that merit function vetoed the OB).  I think that was easy enough to understand and process.  Whichever way we go, I would strongly urge against use of the term ```Weight``` as that would be confusing since one scheduling strategy would multiply each ```Weight``` by a weight before summing the results.

As for allowing an OB to have multiple image sets (i.e. different filters), I think the idea of wanting near simultaneous observations of the same target in different filters in order to measure color is common enough that we need to support that in a straightforward way.

The next step would be linking OBs which is much more complex.  I think that those two are sufficiently different in difficulty and in how commonly used they are that we can support one now and then consider strategies for the other.  I worked with a project which uses the scheduling software that Denny 04 described (now much more advanced) and he does allow for linked OBs which can be quite useful when doing photometric observations, but those require some additional configuration options to manage them, so it would take some careful thought and planning.  In his case, he also handles OBs linked within a night differently from those in separate nights, as to do the latter you need to keep track of what has been observed over many nights, not just within this night.

Lastly, I don't think we need to split out a ```WeightCombiner``` class.  I think we can just add a scheduler class which uses that particular strategy to combine merit values.  So we might end up with a ```SummingScheduler``` to handle weighted sums and another to handle products of merits, etc.  That proliferates several scheduler classes, but I can't see it getting too bad.

**Comment by eteq** [comment]
@joshwalawender - Ok, that mostly sounds good.  Will probably want to create a `BaseScheduler` that abstracts out as much as possible of the bookeeping (e.g. computing slew times, configuration change processing, etc.), so that the different scheduler implementations can focus on just the "weighting" schemes/actual scheduling.  But I definitely agree that it's more conceptually straightforward for the different schemes to have their own scheduler class rather than a hierarchy of weighting classes.

**Comment by eteq** [comment]
Oh, but I am *not* completely sold on the idea of "imagesets" as a distinct entity.  That seems like two ways of specifying the same thing (two OBs vs. two image sets with one OB), and I fear users would intermix the two.  Perhaps the solution is to instead have there be "container" OBs that hold multiple OBs, but are scheduled as a single block?  That way we don't have to build linking into the scheduler itself, but can still stick with "all atomic observations are an OB"?

**Comment by joshwalawender** [comment]
@eteq Perhaps I misunderstand, but I think there is a big difference between two OBs and two image sets with one OB and that would be that one of those must be scheduled together, the other could happen at drastically different times.  I think a project might want the ability to do both and choose to couple or decouple them when defining the observation for the scheduler.

Regarding ImageSet vs. a ContainerOB, either is fine.  I think the differences are primarily semantic.  ImageSet is probably a bad term since that implies it is only for imaging which was not the intent (I stole that terminology from the Scheduler that is described in Denny 2004).  I almost used a name like InstrumentConfig, but that does't imply action or the idea that it occupies time within the OB, so I don't like that either.  

The more I think about it, the more I like your idea of a container OB.  That could be general enough to be used for other things.  For example, if I wanted to couple two different OBs with different targets in to one block that was scheduled together.

**Comment by ejeschke** [comment]
A container (I prefer the term "GroupOB" actually) that could be scheduled as an atomic entity would be really useful.  We are struggling with this right now trying to schedule a calibration target with a science target.  You either want to do both in the same night or not at all.  And such a construct might also be useful for cadence observing over a long-term schedule.

**Comment by joshwalawender** [comment]
Since we hope to use this in project Panoptes I'm more than willing to start writing code, but I think we need to decide on one set of terms.  I'll throw out the following as a starting point:

```ExposureSet``` - Describes a single set up of a particular instrument.  Primarily this is here to hold the configuration and the duration of a particular exposure set.  An instance of this may be reused for several targets when generating ```Observation```s.

```Visit``` - Describes an observation of a single target using a one or more ```ExposureSet```s

```ObservingBlock``` - The minimum schedulable unit.  May contain one or more ```Observation```s.

Handling constraints for OBs which contain visits to multiple targets will be tough, but I think we have to address that for exactly the case @ejeschke describes.

What do you think?

**Comment by bmorris3** [comment]
@joshwalawender In your last comment, when you say: 

> `ObservingBlock` - The minimum schedulable unit. May contain one or more `Observations`.

do you mean "may contain one or more `Visit`s"?

***

@joshwalawender That sounds like a great starting point. I think that's a fairly intuitive combination. 

Would you envision being able to do something like this: 

```python
import astropy.units as u

hour_long_spectrum = ExposureSet(instrument='MOSFIRE', filter='K', duration=1*u.hour)
dim_star_visit = Visit(target=FixedTarget.from_name('GJ 1535'), 
                           exposure_set=hour_long_spectrum)
standard_star_visit = Visit(target=FixedTarget.from_name('Gl 23245'), 
                           exposure_set=hour_long_spectrum)
ob = ObservingBlock(visits=[dim_star_visit, standard_star_visit])
```

***

> Handling constraints for OBs which contain visits to multiple targets will be tough

I don't think it will be so terrible. In order for one OB to be observable, the constraints just have to applied to each target within the OB and reduced with logical ANDs – for a tutorial on how this looks, check out astropy/astroplan#135, particularly [this notebook](http://nbviewer.ipython.org/github/bmorris3/gsoc2015/blob/master/constraints_boolean_logic.ipynb).

**Comment by joshwalawender** [comment]
@bmorris3 Oops, yes, that should have read ```Visit```s.  Also, your example is precisely what I was thinking, I should have thought to include an example like that.

**Comment by bmorris3** [comment]
Great, @joshwalawender. If you have any questions as you start working, don't hesitate to reach out! I'm happy to help bring you up to speed on astroplan, and always interested to see if you discover weaknesses in the [docs](http://astroplan.readthedocs.org).

**Comment by cdeil** [comment]
@eteq - Could you please rebase this and summarise the discussion?
Is it clear what needs to be done before this can go in or is more discussion / feedback needed?

**Comment by eteq** [comment]
#151 supersedes this, so I'm closing it.  I don't know that I can make a useful summary of the discussion because it's fairly wide-ranging, so I'll leave it at that, but we can discuss more in #151

**Pull request closed without merging**